File: cpp/01-getting-started/ex01_roi_test.cpp
/**
 * @file ex01_roi_test.cpp
 * @brief Test 06: Regions of Interest (ROI) and Zero-Copy Matrix Slicing (`cv::Rect`)
 *
 * @par Purpose
 * Demonstrates memory-efficient sub-matrix extraction and manipulation using header referencing (`cv::Rect`).
 * Isolates a rectangular sub-region within a parent matrix without duplicating underlying pixel buffer memory,
 * applies an in-place color transformation to the ROI, and verifies that parent data references update instantly.
 *
 * @author Matha Goram
 * @version 1.0.0
 * @date 2026-08-12
 *
 * @par Copyright
 * Copyright (c) 2026 ParkCircus Productions. All rights reserved.
 *
 * @par MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * @par Update History
 * - v1.0.0 (2026-08-11): Initial implementation with zero-copy sub-matrix ROI slicing and modification.
 *
 * @par Prerequisites
 * - C++17 compliant compiler (GCC / Clang)
 * - CMake 3.22+
 * - OpenCV 4.x or 5.x development libraries (`libopencv-dev`)
 * - Read/Write permissions within the active working build directory
 *
 * @par User Interface Guide
 * Non-interactive Command-Line Interface (CLI) tailored for headless environments:
 * 1. Build the CMake target `ex01_roi_test` in CLion or terminal.
 * 2. Execute the binary on the target node.
 * 3. Inspect telemetry log output in standard console streams (`stdout` / `stderr`).
 * 4. Open `output_roi.png` in CLion's built-in image viewer.
 *
 * @par Processing Workflow and Algorithms
 * 1. **Canvas Allocation**: Instantiate 640x480 neutral background matrix.
 * 2. **ROI Slicing**: Define a `cv::Rect(160, 120, 320, 240)` and extract sub-matrix reference `cv::Mat roi = parent(region)`.
 * 3. **In-Place Modification**: Apply a solid color fill or pattern directly to the `roi` handle.
 * 4. **Frame Serialization**: Export modified parent matrix buffer to PNG image on disk via `cv::imwrite()`.
 * 5. **Error Validation**: Evaluate disk write boolean return code and report execution status.
 *
 * @par Error Message Responses
 * - `[Ex 01 - ROI Slicing] ❌ ERROR: Failed to write output_roi.png`
 *   - **Cause**: Target directory is read-only, disk space exhausted, or invalid target path.
 *   - **Resolution**: Verify directory permissions and disk capacity on target device.
 *
 * @par References and Notes
 * - OpenCV Basic Structures: https://docs.opencv.org/master/d3/d63/classcv_1_1Mat.html
 * - Memory Note: Slicing via `cv::Rect` creates a header pointing to the parent buffer without copying pixel data arrays.
 */

#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>
#include <cstdlib>

int main() {
    const std::string ex_prefix = "Ex 01 - ROI Slicing";
    const std::string output_filename = "output_roi.png";

    std::cout << "[" << ex_prefix << "] 📐 Initializing Region of Interest (ROI) Test...\n";

    // Create a background canvas
    cv::Mat canvas(480, 640, CV_8UC3, cv::Scalar(60, 60, 60));

    // Define a region of interest using cv::Rect(x, y, width, height)
    cv::Rect roi_region(160, 120, 320, 240);

    // Extract sub-matrix (zero-copy header reference)
    cv::Mat roi = canvas(roi_region);

    // Modify the ROI directly (fill with bright orange in BGR)
    roi.setTo(cv::Scalar(0, 128, 255));

    // Draw an inner marker to prove reference connection
    cv::rectangle(roi, cv::Point(40, 40), cv::Point(280, 200), cv::Scalar(255, 255, 255), 3);

    if (!cv::imwrite(output_filename, canvas)) {
        std::cerr << "[" << ex_prefix << "] ❌ ERROR: Failed to write " << output_filename << "\n";
        return EXIT_FAILURE;
    }

    std::cout << "[" << ex_prefix << "] ✅ ROI modified parent frame serialized to: " << output_filename << "\n";
    return EXIT_SUCCESS;
}

File: python/01-getting-started/ex01_roi_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
@file ex01_roi_test.py
@brief Exercise 07: Zero-Copy Memory Slicing and Region of Interest (ROI) Manipulation

@version 1.0.0
@date 2026-08-16
@author Matha Goram (Systems Architect)

@copyright MIT License
Copyright (c) 2026 ParkCircus Productions. All Rights Reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

@par Purpose
Demonstrates extracting and modifying sub-matrices (Regions of Interest) without allocating
duplicate memory buffers, utilizing NumPy slice references.

@par Update History
- v1.0.0 (2026-08-16): Initial Python port for Chapter 01 getting started.

@par Processing Workflow and Algorithms
1. Load source image matrix.
2. Slice target region coordinates using array slicing (`img[y1:y2, x1:x2]`).
3. Apply in-place color transformation to the ROI view.
4. Verify parent matrix modification and write artifact to disk.
"""

import sys
import os
import numpy as np
import cv2

def main():
    ex_prefix = "Ex 01 - ROI"
    input_filename = "output_canvas.png"

    print(f"[{ex_prefix}] 📐 Extracting Region of Interest (ROI) via view slicing...")

    if not os.path.exists(input_filename):
        dummy = np.full((300, 300, 3), (100, 100, 100), dtype=np.uint8)
        cv2.imwrite(input_filename, dummy)

    src = cv2.imread(input_filename, cv2.IMREAD_COLOR)
    if src is None:
        print(f"[{ex_prefix}] ❌ ERROR: Could not load source image.", file=sys.stderr)
        sys.exit(1)

    # Define ROI slice bounds (y: 100 to 250, x: 150 to 350)
    roi = src[100:250, 150:350]

    # In-place color tinting on ROI view (adding green tint)
    roi[:, :, 1] = cv2.add(roi[:, :, 1], 50)

    output_filename = "output_roi_modified.png"
    if not cv2.imwrite(output_filename, src):
        print(f"[{ex_prefix}] ❌ ERROR: Failed to save ROI modified image.", file=sys.stderr)
        sys.exit(1)

    print(f"[{ex_prefix}] ✅ ROI view extraction and in-place modification complete. Saved to {output_filename}.")
    sys.exit(0)

if __name__ == "__main__":
    main()
